lib/Mojo/Redis/Protocol.pm

```diff
@@ -0,0 +1,240 @@
+        $str .= "\$-1\r\n";
+      }
+    }
+    elsif ($type eq '*') {
+      if (defined $data) {
+        $str .= '*' . scalar(@$data) . "\r\n";
+        unshift @stack, @$data;
+      }
+      else {
+        $str .= "*-1\r\n";
+      }
+    }
+    else {
+      confess "[Mojo::Redis::Protocol] Unknown message type $type";
+    }
+  }
+
+  return $str;
+}
+
+sub encoding {
+  my $self = shift;
+  return $self->{encoding} unless @_;
+  $self->{encoding} = shift // '';
+  $self->{encoder} = $self->{encoding} && find_encoding($self->{encoding});
+  Carp::confess("Unknown encoding '$self->{encoding}'") if $self->{encoding} and !$self->{encoder};
+  return $self;
+}
+
+sub new {
+  my $self = shift->SUPER::new(@_);
+  $self->encoding($self->{encoding});
+  @$self{qw(buf len stop)} = ('', -1, -1);
+  return $self;
+}
+
+sub parse {
+  my $self = shift;
+  $self->{buf} .= shift;
+
+  my $curr = $self->{curr} ||= {level => 0};
+  my $buf  = \$self->{buf};
+  my ($encoder, @messages);
+
+CHUNK:
+  while (length $$buf) {
+    do { local $_ = $$buf; s!\r\n!\\r\\n!g; warn "[Mojo::Redis::Protocol] parse($_)\n" } if DEBUG > 1;
+
+    # Look for initial type
+    if (!$curr->{type}) {
+      @$self{qw(len stop)} = (-1, -1);
+      $curr->{type} = substr $$buf, 0, 1, '';
+      _parse_debug(start => $curr) if DEBUG > 1;
+    }
+
+    # Simple Strings, Errors, Integers
+    if ($curr->{type} eq '+' or $curr->{type} eq '-' or $curr->{type} eq ':') {
+      $self->{len} = index $$buf, "\r\n" if $self->{len} < 0;
+      last CHUNK if $self->{len} < 0;    # Wait for more data
+
+      $curr->{data} = substr $$buf, 0, $self->{len}, '';
+      substr $$buf, 0, 2, '';            # Remove \r\n after data
+
+      # Force into correct data type
+      if ($curr->{type} eq ':') {
+        $curr->{data} += 0;
+      }
+      elsif ($encoder ||= $self->{encoder}) {
+        $curr->{data} = $encoder->decode($curr->{data}, 0);
+      }
+
+      _parse_debug(scalar => $curr) if DEBUG > 1;
+    }
+
+    # Bulk Strings
+    elsif ($curr->{type} eq '$') {
+      $self->{stop} = index $$buf, "\r\n" if $self->{stop} < 0;
+      last CHUNK if $self->{stop} < 0;    # Wait for more data
+
+      $self->{len} = substr $$buf, 0, $self->{stop}, '';
+      last CHUNK if length($$buf) - 2 < $self->{len};    # Wait for more data
+
+      $curr->{data} = $self->{len} < 0 ? undef : substr $$buf, 2, $self->{len}, '';
+      $curr->{data} = $encoder->decode($curr->{data}, 0) if $encoder ||= $self->{encoder};
+      substr $$buf, 0, 4, '';                            # Remove \r\n before and after data
+      _parse_debug(scalar => $curr) if DEBUG > 1;
+    }
+
+    # Arrays
+    elsif ($curr->{type} eq '*') {
+      $self->{stop} = index $$buf, "\r\n" if $self->{stop} < 0;
+      last CHUNK if $self->{stop} < 0;                   # Wait for more data
+
+      $curr->{len} = substr $$buf, 0, $self->{stop}, '';
+      last CHUNK if length($$buf) - 2 < $curr->{len};    # Wait for more data
```

Comment:
I don't think this line should be here. The size of the array does not indicate anything about how much data is required.